File: internal/storage/get.go
package storage

import (
	"dvs/internal/file"
	"dvs/internal/log"
	"dvs/internal/meta"
	"dvs/internal/utils"
	"errors"
	"fmt"
	"os"
	"time"
)

// Gets a file from storage
func Get(localPath string, storageDir string, gitDir string, dry bool) error {
	/// Get metadata of specified file
	metadata, err := meta.Load(localPath)
	if os.IsNotExist(err) {
		return errors.New("no metadata")
	} else if err != nil {
		return errors.New("failed to load metadata")
	}

	// Get storage path
	storagePath := getStoragePath(storageDir, metadata.FileHash)

	// Check if file is already present locally
	_, err = os.Stat(localPath)
	var localHash string
	if err == nil {
		// Get local file's hash
		log.Print("    Getting local hash...")

		startTime := time.Now()
		localHash, err = file.GetFileHash(localPath)
		endTime := time.Now()

		log.OverwritePreviousLine()
		if err != nil {
			log.Print("    Getting local hash...", log.ColorBold(log.ColorYellow("!")))
		} else {
			log.Print("    Getting local hash...", log.ColorGreen("✔ in ", utils.FormatDuration(endTime.Sub(startTime))))
		}
	}

	// Copy file to destination
	// if the destination already exists and hash matched, skip copying
	if os.IsNotExist(err) || metadata.FileHash == "" || localHash == "" || metadata.FileHash != localHash {
		err = Copy(storagePath, localPath, dry)
		if err != nil {
			return errors.New(fmt.Sprint("failed to copy file:", err))
		}

		log.OverwritePreviousLine()
		log.Print("    Cleaning up...", log.ColorGreen("✔\n"))
	} else {
		log.Print("    Copying...", log.ColorGreen("✔ file already up to date\n"))
	}

	return nil
}

File: src/internal/storage/get.rs
use std::path::PathBuf;
use anyhow::{Context, Result};
use crate::internal::file::hash;
use crate::internal::meta::file;
use crate::internal::storage::copy;

// gets a file from storage
pub fn get(local_path: &PathBuf, storage_dir: &PathBuf) -> Result<()> {
    // get metadata
    let metadata = file::load(&local_path).with_context(|| format!("could not get metadata file for {}", local_path.display()))?;

    // get storage data
    let storage_path = hash::get_storage_path(storage_dir, &metadata.file_hash);

    // get hashes to compare
    let local_hash = hash::get_file_hash(&local_path);
    let metadata_hash = metadata.file_hash;

    // check if up-to-date file is already present locally
    if !local_path.exists() || local_hash == String::from("") || metadata_hash == String::from("") || local_hash != metadata_hash {
        copy::copy(&storage_path, &local_path).with_context(|| format!("could not copy {} from storage directory: {}", local_path.display(), storage_dir.display()))?;
    }
    else {
        println!("{} already up to date", local_path.display())
    }

    Ok(())
} // get